backends/graphs/controls.cpp

```diff
@@ -107,6 +129,14 @@ bool ControlGraphs::preorder(const IR::PackageBlock *block) {
             controlStack.popBack();
             GraphAttributeSetter()(g_);
             writeGraphToFile(g_, name);
+        } else if (it.second->is<IR::ParserBlock>()) {
```

Comment:
while I'm all for unifying code, it looks like this and the `writeGraphToFile` overload for `P4::ParserCallGraph` do not belong here, not in a class called `ControlGraphs` in any case
I believe it would be better to leave the code for parsers in its own file. It doesn't look that it is re-using a lot of the existing code from `ControlGraphs`which is mostly written around the boost graph library.

Reply:
The problem is that the preorder(IR::Package) method does not belong to the ControlGraphs file itself.

In fact, there is hardly any code used for the parser part.  If you want labels on the edges I will have to duplicate some of the code which I reused, because the ParserControlFlowGraph class does not have edge information.

I could not really find a benefit for using that boost library, it looks overly complicated. The dot file format has been stable for 20 years now, and it is really very simple.

Reply:
With the removal of the old parser code, the current `Graphs` base class doesn't make sense any more. I believe it was just an attempt at unification based when the generation code was similar for parsers and controls. With your current proposal I'm not sure I see the need for a base class at all, or in any case a base class with boost members creeping in.
I don't see why you tried to piggyback the parser graph generation on top of the `ControlGraph` visitor by reusing the same visitor function. It looks like the parser graph generation can be done with its own, very simple, visitor in a cleaner way.